custom_components/mqtt_vacuum_camera/camera.py

```diff
@@ -236,6 +241,134 @@
+                    else:
+                        _LOGGER.debug("No nearby obstacle found.")
+                        self._should_poll = True  # Turn on polling
+                        self._shared.camera_mode = CameraModes.MAP_VIEW
+        else:
+            _LOGGER.debug("No obstacles data available.")
+            self._should_poll = True
+
+    @staticmethod
+    async def _async_find_nearest_obstacle(x, y, obstacles):
+        """Find the nearest obstacle to the given coordinates."""
+        nearest_obstacle = None
+        min_distance = float('inf')  # Start with a very large distance
+        _LOGGER.debug(f"Finding the nearest {min_distance} obstacle to coordinates: {x}, {y}")
+
+        for obstacle in obstacles:
+            obstacle_point = obstacle["point"]
+            obstacle_x = obstacle_point["x"]
+            obstacle_y = obstacle_point["y"]
+
+            # Calculate Euclidean distance
+            distance = math.sqrt((x - obstacle_x) ** 2 + (y - obstacle_y) ** 2)
+
+            if distance < min_distance:
+                min_distance = distance
+                nearest_obstacle = obstacle
+
+        return nearest_obstacle
+
+
+    @staticmethod
+    async def download_image(url: str, storage_path: str, filename: str):
+        """
+        Asynchronously download an image using threading to avoid blocking.
+
+        Args:
+            url (str): The URL to download the image from.
+            storage_path (str): The directory to save the image.
+            filename (str): The name to save the image as.
+
+        Returns:
+            str: The full path to the saved image or None if the download fails.
+        """
+        # Ensure the storage path exists
+        os.makedirs(storage_path, exist_ok=True)
+
+        obstacle_file = os.path.join(storage_path, filename)
+
+        async def blocking_download():
+            """Run the blocking download in a separate thread."""
+            try:
+                async with aiohttp.ClientSession() as session:
+                    async with session.get(url) as response:
+                        if response.status == 200:
+                            with open(obstacle_file, "wb") as f:
+                                f.write(await response.read())
+                            _LOGGER.debug(f"Image downloaded successfully: {obstacle_file}")
+                            return obstacle_file
+                        else:
+                            _LOGGER.warning(f"Failed to download image: {response.status}")
+                            return None
+            except Exception as e:
+                _LOGGER.error(f"Error downloading image: {e}")
+                return None
+
+
+        executor = ThreadPoolExecutor(max_workers=3)  # Limit to 3 workers
+
+        # Run the blocking I/O in a thread
+        return await asyncio.get_running_loop().run_in_executor(executor, asyncio.run, blocking_download())
+
```

Comment:
@coderabbitai we need a thread to read it to avoid blocking the main thread of Home Assistant, as this is a file operation. I tested the code in Home Assistant supervised and I'm going to release this a beta without changes will work on the changes you proposed in a phase 3 soon